### PR TITLE
Make route reference and error logs more robust in the viewport data source

### PR DESCRIPTION
**Pull request by LukasPaczos**
### Description

Closes https://github.com/mapbox/mapbox-navigation-android/issues/4289.

This PR makes stronger checks against the route reference and aborts early if there's a difference between what is provided in `#onRouteChanged` and `#onRouteProgressChanged`.

I initially hesitated to use the `isSameRoute` extension in fear of poor performance for longer routes and the frequency of invocations, but after measuring on Pixel 2XL when most checks fail on geometry string comparison:
- regular routes (sub 100 miles) take ~0.01 of a millisecond to compare
- long routes (1k miles) take ~0.1 of a millisecond to compare
- while insanely long route (Europe to India) take ~0.3 of a millisecond to compare

Given that performance, I think it's safe to call it more often and make the developer experience better.

### Changelog

```
<changelog>Made route reference checks more robust in the `MapboxNavigationViewportDataSource` and improved logs to avoid situations where a route reference is mismatched and could produce incorrect frames or crash.</changelog>
```

### CODE REVIEW

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/camera/data/MapboxNavigationViewportDataSource.kt [hunk 1]
```diff
@@ -434,16 +433,27 @@ class MapboxNavigationViewportDataSource(
...
-                Message("Provided route and navigated route do not have the same UUID.")
+                Message(
+                    "You're calling #onRouteProgressChanged but you didn't call #onRouteChanged."
+                )
             )
+            clearProgressData()
+            return
+        } else if (!currentRoute.isSameRoute(routeProgress.route)) {
+            LoggerProvider.logger.e(
+                Tag("MbxViewportDataSource"),
```
Comment (Guardiola31337) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/camera/data/MapboxNavigationViewportDataSource.kt:448]: Could we extract this magic number into a `const`ant?

```suggestion
                Tag(TAG),
```

File: libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/camera/data/MapboxNavigationViewportDataSourceTest.kt [hunk 2]
```diff
@@ -949,6 +955,153 @@ class MapboxNavigationViewportDataSourceTest {
...
+        }
+        val legProgress = mockk<RouteLegProgress> {
+            every { currentStepProgress } returns stepProgress
+        }
+        every { routeProgress.currentLegProgress } returns legProgress
+        every { routeProgress.route } returns route
+
+        val location = createLocation()
+
+        // following mocks
```
Comment (Guardiola31337) [inline libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/camera/data/MapboxNavigationViewportDataSourceTest.kt:981]: Could we name the variable differently (e.g. `mockedFollowingCameraOptions`) and remove the comment?

The same thing applies to other comments 👇

**Review by Guardiola31337 — APPROVED** [review]
Left a couple of minor comments.

File: libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/camera/data/MapboxNavigationViewportDataSourceTest.kt (hunk 2 above)
Comment (LukasPaczos) [reply libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/camera/data/MapboxNavigationViewportDataSourceTest.kt:981]: That wasn't the goal, it was just to visually structure the code a little bit.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/camera/data/MapboxNavigationViewportDataSource.kt (hunk 1 above)
Comment (LukasPaczos) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/camera/data/MapboxNavigationViewportDataSource.kt:448]: I'll extract it :+1:

**Pull request merged**